cores/esp8266/core_esp8266_postmortem.cpp

```diff
@@ -322,5 +403,199 @@ void __stack_chk_fail(void) {
+    "ill\n\t"
+    ".size postmortem_double_exception_handler, .-postmortem_double_exception_handler\n\t"
+);
+
+/*
+  Of the Exception Cause table's 64 entries, most are set to
+  `_xtos_unhandled_exception`. And I assume most of these never occur; however,
+  at lesat one does exccause 20. This one occurs on a call to a NULL pointer. This
+  can happen when a function prototype is given the weak attribute but never
+  defined, or forgetting to test a call back function pointer for NULL before
+  calling, etc.
+
+  A problem with the `_xtos_unhandled_exception` function is that it handles the
+  event with a Debug Exception BP. If you don't have GDB running you see a HWDT
+  reset. If interrupts are at INTLEVEL 2 or higher (a BP instruction becomes a
+  NOP), you still see a HWDT reset regardless of running GDB.
+*/
+typedef void (* _xtos_handler)(void);
+static  _xtos_handler * const _xtos_exc_handler_table = (_xtos_handler *)0x3FFFC000u;
+#define ROM_xtos_unhandled_exception (reinterpret_cast<_xtos_handler>(0x4000dc44))
+
+void postmortem_xtos_unhandled_exception(void);
+asm(
+    ".section     .iram.text.postmortem_xtos_unhandled_exception,\"ax\",@progbits\n\t"
+    ".literal_position\n\t"
+    ".align       4\n\t"
+    ".type        postmortem_xtos_unhandled_exception, @function\n\t"
+    "\n"
+"postmortem_xtos_unhandled_exception:\n\t"
+    // Rewind Boot ROM User Exception prologue and continue to
+    // _KernelExceptionVector such that we look a Debug BP Exception sitting at
+    // 'break 1, 1' in the _xtos_unhandled_exception handler.
+    "movi         a2,     0x4000dc4bu\n\t"
+    "wsr.epc2     a2\n\t"
+    "l32i         a2,     a1,     20\n\t"
+    "l32i         a3,     a1,     24\n\t"
+    "addmi        a1,     a1,     0x100\n\t"
+    "j            _KernelExceptionVector\n\t"
+    ".size postmortem_xtos_unhandled_exception, .-postmortem_xtos_unhandled_exception\n\t"
+);
+
+
+static void replace_exception_handler_on_match(
+                uint32_t cause,
+                _xtos_handler match,
+                _xtos_handler replacement) {
+
+    _xtos_handler old_wrapper = _xtos_exc_handler_table[cause];
+    if (old_wrapper == match || NULL == match) {
+        _xtos_exc_handler_table[cause] = replacement;
+    }
+}
+
+static void install_unhandled_exception_handler(void) {
+    // Only replace Exception Table entries still using the orignal Boot ROM
+    // _xtos_unhandled_exception handler.
+    for (size_t i = 0; i < 64; i++) {
+        replace_exception_handler_on_match(
+            i,
+            ROM_xtos_unhandled_exception,
+            (_xtos_handler)postmortem_xtos_unhandled_exception);
+    }
+}
+
+void postmortem_init(void) {
+    if (!gdb_present()) {
+        // Install all three (interdependent) exception handler stubs.
+        // Length of copy must be rounded up to copy memory in aligned 4 byte
+        // increments to comply with IRAM memory access requirements.
+        uint32_t save_ps = xt_rsil(15);
```

Comment:
What do 16 and 32 mean? doea it make sense to move to a define. constexpr and comments?

Reply:
I initially wasn't sure how to generate values at compile time from the ASM blocks sizes. I think I have that resolved.